src/number_theory/multiplicity.lean

```diff
@@ -0,0 +1,454 @@
+/-
+Copyright (c) 2022 Tian Chen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tian Chen
+-/
+import algebra.geom_sum
+import data.int.parity
+import number_theory.padics.padic_norm
+
+/-!
+# Multiplicity in Number Theory
+
+This file contains results in number theory relating to multiplicity.
+
+## Main statements
+
+* `multiplicity.int.pow_sub_pow` is the lifting the exponent lemma for odd primes.
+  We also prove several variations of the lemma.
+
+## References
+
+* [Wikipedia, *Lifting-the-exponent lemma*]
+  (https://en.wikipedia.org/wiki/Lifting-the-exponent_lemma)
+-/
+
+open ideal ideal.quotient
+open_locale big_operators
+
+variables {R : Type*} {n : ℕ}
+
+
+section comm_ring
+variables [comm_ring R] {a b x y : R}
+
+
+lemma dvd_geom_sum₂ {x y : R} (h : ↑n ∣ x - y) : ↑n ∣ geom_sum₂ x y n :=
+begin
+  rw ← mem_span_singleton at ⊢ h,
+  rw ← ideal.quotient.eq at h,
+  rw [← eq_zero_iff_mem, ring_hom.map_geom_sum₂, h, geom_sum₂_self],
+  apply mul_eq_zero_of_left,
+  rw [← map_nat_cast (mk $ span ({n} : set R)) n, eq_zero_iff_mem, mem_span_singleton]
+end
+
+lemma sq_dvd_add_mul_pow_sub (p x y : R) (n : ℕ) :
+  p ^ 2 ∣ (x + p * y) ^ n - (x ^ (n - 1) * (p * y) * n + x ^ n) :=
+begin
+  cases n,
+  { simp only [pow_zero, nat.cast_zero, mul_zero, sub_self, dvd_zero, zero_add] },
+  { simp only [nat.succ_sub_succ_eq_sub, tsub_zero, nat.cast_succ, add_pow,
+    finset.sum_range_succ, nat.choose_self, nat.succ_sub _, tsub_self, pow_one,
+    nat.choose_succ_self_right, pow_zero, mul_one, nat.cast_zero, zero_add, nat.succ_eq_add_one],
+    abel,
+    apply finset.dvd_sum,
+    intros x hx,
+    apply dvd_mul_of_dvd_left,
+    apply dvd_mul_of_dvd_right,
+    rw mul_pow,
+    apply dvd_mul_of_dvd_left,
+    apply pow_dvd_pow,
+    apply le_tsub_of_add_le_left,
+    linarith [finset.mem_range.mp hx]  },
+end
+
+lemma not_dvd_geom_sum₂ {p : R} (hp : prime p)
+  (hxy : p ∣ x - y) (hx : ¬p ∣ x) (hn : ¬p ∣ n) :
+  ¬p ∣ geom_sum₂ x y n :=
+begin
+  rw ← mem_span_singleton at *,
+  rw ← ideal.quotient.eq at hxy,
+  rw ← eq_zero_iff_mem at *,
+  haveI := (span_singleton_prime hp.1).mpr hp,
+  rw [ring_hom.map_geom_sum₂, ←hxy, geom_sum₂_self, mul_eq_zero],
+  exact not_or (by rwa map_nat_cast at hn) (λ h, hx $ pow_eq_zero h)
+end
+
+variables {p : ℕ}
+
+lemma odd_sq_dvd_geom_sum₂_sub (hp : odd p) :
+  ↑p ^ 2 ∣ geom_sum₂ (a + p * b) a p - p * a ^ (p - 1) :=
+begin
```

Comment:
Could you turn this proof into a `calc` block?

Reply:
Turned most of the proof into a calc block in the new commit. Sadly, I think it ended up being longer than the existing proof :/

Reply:
Thanks for the expansion. I like this form more despite the length, because you have some more guidance as you read it.